packages/panzer/adapters-stk/src/evaluators/Panzer_STK_ProjectField_impl.hpp

```diff
@@ -0,0 +1,139 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Roger P. Pawlowski (rppawlo@sandia.gov) and
+// Eric C. Cyr (eccyr@sandia.gov)
+// ***********************************************************************
+// @HEADER
+
+#ifndef PANZER_STK_PROJECT_FIELD_IMPL_HPP
+#define PANZER_STK_PROJECT_FIELD_IMPL_HPP
+
+#include "Teuchos_Assert.hpp"
+#include "Phalanx_DataLayout.hpp"
+
+#include "Intrepid2_ProjectionTools.hpp"
+#include "Intrepid2_OrientationTools.hpp"
+
+#include "Panzer_PureBasis.hpp"
+#include "Panzer_HierarchicParallelism.hpp"
+#include "Kokkos_ViewFactory.hpp"
+
+#include "Teuchos_FancyOStream.hpp"
+
+namespace panzer_stk {
+
+template<typename EvalT,typename Traits>
+ProjectField<EvalT, Traits>::
+ProjectField(const std::string & inName, Teuchos::RCP<panzer::PureBasis> src,
+             Teuchos::RCP<panzer::PureBasis> dst, 
+             std::string outName):
+  srcBasis_(src), dstBasis_(dst)
+{ 
+  using panzer::Cell;
+  using panzer::BASIS;
+
+  static_assert(std::is_same<EvalT,panzer::Traits::Residual>::value);
+
+  Teuchos::RCP<PHX::DataLayout> srcBasis_layout = srcBasis_->functional;
+  Teuchos::RCP<PHX::DataLayout> dstBasis_layout = dstBasis_->functional;
+
+  if (outName == "") outName = inName; 
+  result_ = PHX::MDField<ScalarT,Cell,BASIS>(outName,dstBasis_layout);
+  this->addEvaluatedField(result_);
+
+  // This shouldn't get modified but needs to be non const 
+  // because of downstream templating in intrepid2
+  source_ = PHX::MDField<ScalarT,Cell,BASIS>(inName,srcBasis_layout);
+  this->addNonConstDependentField(source_);
+
+  this->setName("Project Field");
+
+}
+
+// **********************************************************************
+template<typename EvalT,typename Traits>
+void ProjectField<EvalT, Traits>::
+postRegistrationSetup(typename Traits::SetupData  d, 
+		      PHX::FieldManager<Traits>& /* fm */)
+{
+  orientations_ = d.orientations_;
+}
+
+// **********************************************************************
+template<typename EvalT,typename Traits>
+void ProjectField<EvalT, Traits>::
+evaluateFields(typename Traits::EvalData workset)
+{ 
+
+  // Is there a chance workset is empty?
+  if (workset.num_cells<=0) return;
+
+  // Perform local L2 projection
+
+  typedef Intrepid2::Experimental::ProjectionTools<PHX::Device> pts;
+
+  // First, need to copy orientations to device
+  auto numCells = source_.extent(0);
+  auto orts = Kokkos::DynRankView<Intrepid2::Orientation,PHX::Device>("orts",numCells);
+  auto orts_host = Kokkos::create_mirror_view(orts);
+
+  if (orientations_ == Teuchos::null) {
```

Comment:
I do not think you need this. If the basis does not requires the orientations, it should not matter what values are stored in the orts, so, no need to set it to Intrepid2::Orientation().  
If this is not the case, then there is probably a bug somewhere.

Reply:
Never mind, I didn't look at the code carefully. What you do is fine.

Is there any plan for panzer to store the orientations directly on device?

Reply:
@mperego - we do plan to store the orientations directly on device. Hope to do that next FY...